### ISSUE TITLE
[Feature] Game Console Acceleration/Full cone NAT

### ISSUE THREAD

**Issue by ghost** [Mygod/VPNHotspot#130]
Inspired by [NetchX/Netch](https://github.com/NetchX/Netch). May be a duplicate of #25.

Is it possible to proxy all UDP traffic from a game console and fix NAT type?

**Comment by Mygod** [comment]
This app already proxy all UDP traffic at its current state. NAT type depends on the VPN you are using.

**Comment by aka-rin** [comment]
Thanks for your reply. I have tested it with my Nintendo Switch console.

It turns out that:

- If I use WireGuard/AnyConnect/... with VPN Hotspot => NAT Type B
- If I use V2rayNG/SSR with UDP forward enabled and VPN Hotspot => NAT Type D
- If I use Kitsunebi/SSR with UDP forward disabled and VPN Hotspot => NAT Type F

According to a thread on Reddit, NAT Type A might be a full cone NAT, Type B might be an address-restricted cone NAT, Type C might be a port-restricted cone NAT and Type D might be a symmetric NAT. Type A or Type B is recommended for online games since they might use different ports while connecting.

Although WireGuard/AnyConnect/... has a Type B, it's still not stable to play online. V2ray/... might have a better experience, so is it possible to fix their NAT types? Thanks in advance.

By the way, I also tested a [game accelerator](http://mqjsq.com) and it did return NAT Type A. I'm just curious how it works...

**Comment by Mygod** [comment]
So from the sound of it, your NAT type is restricted by what VPN you are using, and VPN Hotspot in principle supports every NAT type. This seems like a good news to me.

NAT types on game consoles are outside of my expertise (and iirc they are documented very poorly). If there are improvements that can be made for this app, I am happy to learn about it; if it is more regarding the VPN app itself, there is nothing I can do over here.

**Comment by aka-rin** [comment]
According to the thread on Reddit, Nintendo use [STUN](https://zh.wikipedia.org/wiki/STUN) to probe your NAT type. Here is the diagram:

![STUN Algorithm](https://zh.wikipedia.org/wiki/STUN#/media/File:STUN_Algorithm3.svg)

So I guess:

- NAT Type F means all UDP connections to STUN servers have been blocked.
- NAT Type D means you have dynamic public IPs or ports while connecting to different STUN servers. It's probably caused by port randomization. [Link](https://www.reddit.com/r/NintendoSwitch/comments/amn7e9/finally_fixed_my_nat_type_d_issue/)
- NAT Type C means the STUN server you are connecting cannot send message to you via a different server port. NAT Type B means another STUN server can't send message to you before you connect to it. According to some pages, it can be fixed by assigning a static IP to game console and [DMZ](https://en.wikipedia.org/wiki/DMZ_(computing)#DMZ_host)(ports forward) enabled on a router, though it may increase security risk.

Can static IP/port assignment and DMZ be implemented in VPN Hotspot?

**Comment by Mygod** [comment]
Looks like Android only supports symmetric NAT. I guess I will keep this open for now but I personally have no plans for investigating further.

**Comment by KeyuanHuang** [comment]
> Looks like Android only supports symmetric NAT. I guess I will keep this open for now but I personally have no plans for investigating further.

I think there should be some way to support fullcone NAT because I've seen some game boosters can support fullcone NAT. But I don't know how it works. For example, I use [Lingti (in Chinese is 灵缇) game booster](https://lingti.com/) on Android to proxy my Nintendo Switch and it could optimize the NAT type to A even without root. The game booster works by connecting the Android and Switch to the same WiFi and then setting the Android as the gateway of Nintendo Switch.

**Comment by Mygod** [comment]
@KeyuanHuang At a quick glance, that app relies on using a proxy server. If you are okay with configuring a proxy server on all of your clients, you can achieve the same effect by using some proxy server app on your phone.

**Comment by KeyuanHuang** [comment]
> @KeyuanHuang At a quick glance, that app relies on using a proxy server. If you are okay with configuring a proxy server on all of your clients, you can achieve the same effect by using some proxy server app on your phone.

But [as mentioned above in this issue](https://github.com/Mygod/VPNHotspot/issues/130#issuecomment-541820210), there's always some NAT type problems with VPN Hotspot on Android. I have several remote proxy servers that support fullcone NAT and I can get fullcone NAT using other apps, such as the TUN mode of [Clash for Windows](https://github.com/Fndroid/clash_for_windows_pkg). It works by generating a virtual network adapter and sharing it via Windows hotspot, which seems very similar to how VPN Hotspot works. But when I connect to the same proxy servers using [Clash for Android](https://github.com/Kr328/ClashForAndroid) and then share them using the repeater mode of VPN Hotspot, the NAT type is symmetric. In this case, it seems that the NAT problems are not caused by proxy servers. 

Therefore, the problems should be caused by other factors. As I mentioned before, some game boosters on Android can achieve fullcone NAT while working as gateways. This suggests that it's possible to get fullcone NAT on Android. I'm wondering whether fullcone NAT is possible for VPN Hotspot. From the observations mentioned above, I believe it's possible, though I don't know how to (I don't know how to code Android apps). If you could exploit this possibility, I'll appreciate it a lot.

**Comment by Mygod** [comment]
I meant "proxy server" running on your Android device. Sorry for the confusion. It seems that Lingti requires you to configure proxy server on your clients (see their instructions).

**Comment by KeyuanHuang** [comment]
> I meant "proxy server" running on your Android device. Sorry for the confusion. It seems that Lingti requires you to configure proxy server on your clients (see their instructions).

You're right. The proxy app will cause some difference. I test Clash for Magisk recently and the NAT type is B. 
Do you know any Android proxy app supporting fullcone NAT?

**Comment by Mygod** [comment]
Investigated on a rooted Pixel running Android 16 (CP1A.260305.018), FULLCONENAT is not available as an iptables target. Possible paths forward:

- Experimental kernel NAT tweak: try a less-random / more deterministic SNAT mode, but benefit is uncertain and it would not provide true full-cone NAT.
- Userspace UDP relay/proxy: likely the only way to approximate fuller cone behavior without extra kernel support, but this is a much larger feature with significant overhead.
- Kernel module / custom kernel path: add something like xt_FULLCONENAT or equivalent, which is the clearest technical route to true full-cone behavior, but it is outside normal app scope and would likely not be portable.

**Comment by wzhy90** [comment]
> Investigated on a rooted Pixel running Android 16 (CP1A.260305.018), FULLCONENAT is not available as an iptables target. Possible paths forward:
> 
> * Experimental kernel NAT tweak: try a less-random / more deterministic SNAT mode, but benefit is uncertain and it would not provide true full-cone NAT.
> * Userspace UDP relay/proxy: likely the only way to approximate fuller cone behavior without extra kernel support, but this is a much larger feature with significant overhead.
> * Kernel module / custom kernel path: add something like xt_FULLCONENAT or equivalent, which is the clearest technical route to true full-cone behavior, but it is outside normal app scope and would likely not be portable.

By using AI to analyze the `iptables-save` rules, I have successfully got NAT A on the Switch.
First of all, add switch's IP address as an exception to masquerading using `-s 192.168.49.xx/32`. Then add rules in FORWARD table to accept switch 's IP address
`-I FORWARD -i tun0 -o p2p0 -d 192.168.49.xx -j ACCEPT
iptables -I FORWARD -i p2p0 -o tun0 -s 192.168.49.xx -j ACCEPT`
After these, you can get Nat A on your switch

 I'm using Android 13 `Linux localhost 5.15.74-android13-8-04992-g623fe372881f-ab10534577 #1 SMP PREEMPT Thu Jul 20 07:56:12 UTC 2023 aarch64 Toybox`

**Comment by Mygod** [comment]
@wzhy90 Thanks, but this is effectively DMZ-style forwarding, not full-cone NAT. The rules allow all forwarded traffic to/from the Switch, while excluding it from `MASQUERADE` means any NAT mapping is handled elsewhere, likely by the VPN app or server.
